---
title: "Detecting doublets in scRNAseq"
author: "Pierre-Luc Germain"
date: "5/4/2019"
output:
  html_document:
    toc: true
    theme: 'cerulean'
    highlight: 'tango'
    code_folding: hide
    keep_md: yes
---

```{r}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(scran)
  library(igraph)
  library(ROCit)
  library(ggplot2)
  library(cowplot)
  library(plDoubletFinder)
})
```

This is an attempt at recreating the rough logic of [DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder), which creates random artificial doublets to detect real doublets, but in a more efficient manner (i.e. simpler, faster, and apparently more accurate). The main changes are:

* we work directly on a reduced count matrix (using rank correlation on top expressed genes, a la [scran](https://github.com/MarioniLab/scran) ), making the detection independent of downstream processing. This avoids the need for all the Seurat pre-processing, making the detection faster and compatible with other analytic choices.
* similarly to `scran` normalization, we use `scran::quickCluster` to first partition the cells into rough clusters, for a fraction of the doublets with prioritize pairs of cells across clusters. We also produce meta-cells from clusters, and create doublets and triplets from them. This strategy enables the detection of most doublets with much fewer artificial doublets.
* while we also rely on the expected proportion of doublets to threshold the scores, we include a variability in the estimate of the doublet proportion (`dbr.sd`), and use the error rate of the real/artificial predicition in conjunction with the deviation in global doublet rate to set the threshold.

The package is available at https://github.com/plger/plDoubletFinder and relies on [scran](https://github.com/MarioniLab/scran) for various steps. The approach is also very similar to [scran::doubletCells](https://rdrr.io/github/MarioniLab/scran/man/doubletCells.html) but simpler.

We test it on the mixology datasets, where doublets are known from the SNPs (multithreading disabled here).

# The 3 cell lines dataset

```{r}
set.seed(1234)
sce <- readRDS("../comparison/datasets/mixology10x3cl.SCE.rds")
# the command:
system.time(d <- plDoubletFinder(sce, 2000))
# evaluation:
d$truth <- factor(as.character(sce$demuxlet_cls), levels=c("SNG","DBL"))
plot(rocit(d$ratio, d$truth), main="mixology 10x 3 cell lines")
table(truth=d$truth, predicted=d$classification)
```

That's somewhat too easy...

<br/><br/>

# The 5 cell lines dataset

Scran's `quickCluster` at the beginning is actually the time-consuming part; if already ran (e.g. in the context of scran normalization), it can be provided:

```{r}
sce <- readRDS("../comparison/datasets/mixology10x5cl.SCE.rds")
system.time(d <- plDoubletFinder(sce, clusters=sce$quickClusters))
d$truth <- factor(as.character(sce$demuxlet_cls), levels=c("SNG","DBL"))
plot(rocit(d$ratio, d$truth), main="mixology 10x 5 cell lines")
table(truth=d$truth, predicted=d$classification)
```

Here it seems that we miss many doublets. However, it is quite likely that these are wrongly called doublets by `demuxlet`. The demuxlet probability of being a doublet is correlated with the number of SNPs and their coverage:

```{r}
CD <- cbind(d, as.data.frame(colData(sce)))
ggplot(CD, aes(-log10(demuxlet.PRB.DBL), demuxlet.N.SNP, colour=log(demuxlet.uniq.reads))) + geom_point()
```

If we look at demuxlet doublets missed by our approach (lower-left panel below), they tend to have a lower number of SNPs and coverage:
```{r}
ggplot(CD, aes(-log10(demuxlet.PRB.DBL), demuxlet.N.SNP, colour=log10(demuxlet.uniq.reads))) + geom_point() + facet_wrap(classification~demuxlet_cls)
```

A lot of them also do not show the high number of detected features with is so common (and expected) among doublets:
```{r}
p <- ggplot(CD, aes(log10_total_counts, total_features, colour=ratio)) + geom_point() + facet_wrap(~demuxlet_cls)
p + geom_point(data=CD[which(CD$demuxlet_cls=="DBL" & CD$classification=="singlet"),], shape=1, size=4)
```

## Comparison with other methods

We next compare with `DoubletFinder` package, which took considerably longer to run. We followed the example on the vignette, with the exception of the doublet rate, which was set to the real value in this dataset.

The `DoubletFinder` confusion matrix:
```{r}
table(truth=CD$demuxlet_cls, predicted=CD$doubletFinder.class)
```

Since `scran::doubletCells` doesn't threshold, we use the real number of doublets:
```{r}
# reducing the object doesn't affect scran accuracy and considerably speeds up:
sce.red <- sce[order(Matrix::rowMeans(counts(sce)), decreasing=TRUE)[1:3000],]
system.time(scores <- scran::doubletCells(counts(sce.red)))
scran.threshold <- scores[order(scores, decreasing=TRUE)[sum(sce$demuxlet_cls=="DBL")]]
scran.class <- ifelse(scores >= scran.threshold, "doublet", "singlet")
table(truth=CD$demuxlet_cls, predicted=scran.class)
```

We also compare it to [scds](https://github.com/kostkalab/scds) :

```{r}
library(scds)
system.time({
  sce = cxds(sce)
  sce = bcds(sce)
  sce = cxds_bcds_hybrid(sce)
})
scds.threshold <- sce$hybrid_score[order(sce$hybrid_score, decreasing=TRUE)[sum(sce$demuxlet_cls=="DBL")]]
scds.class <- ifelse(sce$hybrid_score >= scds.threshold, "doublet", "singlet")
table(truth=sce$demuxlet_cls, predicted=scds.class)
```

```{r}
truth <- factor(CD$demuxlet_cls, c("SNG","DBL"))
roclist <- lapply( list(pl.ratio=CD$ratio,
                        DoubletFinder.p=CD$doubletFinder.pANN_0.25_0.02_82,
                        scran.score=scores,
                        scds.hybrid=sce$hybrid_score), 
                   class=truth, FUN=rocit)
colors <- c("blue","red","black","pink")
plot(roclist[[1]], lwd=2, col=c(colors[[1]], "grey"), legend=FALSE, YIndex=FALSE)
for(i in 2:length(roclist)){
  lines(roclist[[i]]$FPR, roclist[[i]]$TPR, lwd=2, col=colors[[i]])
}
legend("bottomright", fill=colors, legend=names(roclist))
```

```{r}
sessionInfo()
```